lib/facter/yum_updates.rb

```diff
@@ -0,0 +1,48 @@
+yum_package_updates = nil
+
+# Executes the upgrading of packages
+# @param
+#   command Command passed into yum command arguments i.e. 'update' or 'list updates'
+def get_updates(command)
+  yum_updates = nil
+  if File.executable?('/usr/bin/yum')
+    yum_get_result = Facter::Util::Resolution.exec("/usr/bin/yum -y -q #{command} 2>&1")
+    unless yum_get_result.nil?
+      yum_updates = []
+      first_line = true
+      yum_get_result.each_line do |line|
+        if first_line
+          first_line = false
+          next
+        end
+        package, _available_versiom, _repository = line.split(%r{\s+})
+        yum_updates.push(package)
+      end
+    end
+  end
+
+  yum_updates
+end
+
+yum_package_updates = get_updates('list updates')
```

Comment:
does it download data from repositories or does the command only check against the local db for known updates? The first way might slow down puppet runs significantly.

Reply:
The default behavior of yum is to check if the cache is out-of-date and to update it if necessary.  So this has indeed an impact when the fact is computed and the package repository has been updated.

We can add a `-C` flag here to use local data only.

We can add a new task to refresh the package list, however both `update` and `upgrade` are existing tasks names (both install the latest version of all installed packages; while the apt module use `update` to update the repositories and `upgrade` to install the upgrades).  I do not know if it makes sense to attempt to have a comparable API.

What do you think?

Reply:
Can't you confine to a fact and avoid running this code on every OS? Even better is to have one fact `yum_package_updates` and use that fact in the other facts.

Reply:
That completely make sense… I took the apt facts as a base and was a bit confused by the way it was done…  but moving the code to a fact and using the valie of the facts to build the other ones looks :100:

Reply:
I stuck to the initial idea of not providing `yum_package_updates` and `yum_updates` when no update is available.  If we can cope with an empty array and 0 in such a case, we can simplify things a bit further.  I personally do not care about this :wink:

Reply:
I'd confine it to the RH OS family. Then an empty array and 0 are good and probably even useful.

Reply:
Did that, I updated the PR description accordingly.  Thanks!

Reply:
There was 2 discussions here, and AFAIAC both seems to be resolved:
1. Only using cache data when facts are collected to avoid slowdowns;
2. Improving facts confinment.

Please reopen if I missed something, thanks!